File: src/libs/ikg2p/src/zhg2p.cpp
#include "zhg2p.h"
#include "g2pglobal.h"
#include "zhg2p_p.h"

#include <QDebug>
#include <utility>

namespace IKg2p {

    static const QMap<QChar, QString> numMap = {
        {'0', "零"},
        {'1', "一"},
        {'2', "二"},
        {'3', "三"},
        {'4', "四"},
        {'5', "五"},
        {'6', "六"},
        {'7', "七"},
        {'8', "八"},
        {'9', "九"}
    };

    static QString joinView(const QList<QStringView> &viewList, const QStringView &separator) {
        if (viewList.isEmpty())
            return {};

        QString result;
        for (auto it = viewList.begin(); it != viewList.end() - 1; ++it) {
            result += it->toString();
            result += separator;
        }
        result += (viewList.end() - 1)->toString();
        return result;
    }

    // reset pinyin to raw string
    static QString resetZH(const QList<QStringView> &input, const QList<QStringView> &res,
                           const QList<int> &positions) {
        QList<QStringView> result = input;
        for (int i = 0; i < positions.size(); i++) {
            result.replace(positions[i], res.at(i));
        }
        return joinView(result, QStringLiteral(" "));
    }

    // delete elements from the list
    template <class T>
    static inline void removeElements(QList<T> &list, int start, int n) {
        list.erase(list.begin() + start, list.begin() + start + n);
    }

    ZhG2pPrivate::ZhG2pPrivate(QString language) : m_language(std::move(language)) {
    }

    ZhG2pPrivate::~ZhG2pPrivate() {
    }

    static void copyStringViewHash(const QHash<QString, QString> &src, QHash<QStringView, QStringView> &dest) {
        for (auto it = src.begin(); it != src.end(); ++it) {
            dest.insert(it.key(), it.value());
        }
    }

    static void copyStringViewListHash(const QHash<QString, QStringList> &src,
                                       QHash<QStringView, QList<QStringView>> &dest) {
        for (auto it = src.begin(); it != src.end(); ++it) {
            QList<QStringView> viewList;
            viewList.reserve(it.value().size());
            for (const auto &item : it.value()) {
                viewList.push_back(item);
            }
            dest.insert(it.key(), viewList);
        }
    }

    // load zh convert dict
    void ZhG2pPrivate::init() {
        QString dict_dir = dictionaryPath() + "/" + m_language;

        loadDict(dict_dir, "phrases_map.txt", phrases_map);
        loadDict(dict_dir, "phrases_dict.txt", phrases_dict);
        loadDict(dict_dir, "user_dict.txt", phrases_dict);
        loadDict(dict_dir, "word.txt", word_dict);
        loadDict(dict_dir, "trans_word.txt", trans_dict);

        copyStringViewHash(phrases_map, phrases_map2);
        copyStringViewListHash(phrases_dict, phrases_dict2);
        copyStringViewListHash(word_dict, word_dict2);
        copyStringViewHash(trans_dict, trans_dict2);
    }

    // get all chinese characters and positions in the list
    void ZhG2pPrivate::zhPosition(const QList<QStringView> &input, QList<QStringView> &res, QList<int> &positions,
                                  bool convertNum) const {
        for (int i = 0; i < input.size(); ++i) {
            const auto &item = input.at(i);
            if (item.isEmpty())
                continue;

            if (word_dict2.contains(item) || trans_dict2.contains(item)) {
                res.append(item);
                positions.append(i);
                continue;
            }

            if (!convertNum) {
                continue;
            }

            auto it = numMap.find(input.at(i).front());
            if (it != numMap.end()) {
                res.append(it.value());
                positions.append(i);
            }
        }
    }


    ZhG2p::ZhG2p(QString language, QObject *parent) : ZhG2p(*new ZhG2pPrivate(std::move(language)), parent) {
    }

    ZhG2p::~ZhG2p() {
    }

    QString ZhG2p::convert(const QString &input, bool tone, bool convertNum, errorType error) {
        return convert(splitString(input), tone, convertNum, error);
    }

    QString ZhG2p::convert(const QList<QStringView> &input, bool tone, bool convertNum, errorType error) {
        Q_D(const ZhG2p);
        QList<QStringView> inputList;
        QList<int> inputPos;

        // get char&pos in dict
        d->zhPosition(input, inputList, inputPos, convertNum);

        // Alloc 1
        QString cleanInputRaw = joinView(inputList, QString());
        QStringView cleanInput(cleanInputRaw);

        QList<QStringView> result;
        int cursor = 0;
        while (cursor < inputList.size()) {
            // get char
            const QStringView &raw_current_char = inputList.at(cursor);
            QStringView current_char;
            current_char = d->tradToSim(raw_current_char);

            // not in dict
            if (d->word_dict2.find(current_char) == d->word_dict2.end()) {
                result.append(current_char);
                cursor++;
                continue;
            }

            // is polypropylene
            if (!d->isPolyphonic(current_char)) {
                result.append(d->getDefaultPinyin(current_char));
                cursor++;
            } else {
                bool found = false;
                for (int length = 4; length >= 2; length--) {
                    if (cursor + length <= inputList.size()) {
                        // cursor: 地, subPhrase: 地久天长
                        QStringView sub_phrase = cleanInput.mid(cursor, length);
                        if (d->phrases_dict2.find(sub_phrase) != d->phrases_dict2.end()) {
                            result.append(d->phrases_dict2[sub_phrase]);
                            cursor += length;
                            found = true;
                        }

                        if (found) {
                            break;
                        }

                        if (cursor >= 1) {
                            // cursor: 重, subPhrase_1: 语重心长
                            QStringView sub_phrase_1 = cleanInput.mid(cursor - 1, length);
                            auto it = d->phrases_dict2.find(sub_phrase_1);
                            if (it != d->phrases_dict2.end()) {
                                result.removeAt(result.size() - 1);
                                result.append(it.value());
                                cursor += length - 1;
                                found = true;
                            }
                        }
                    }

                    if (found) {
                        break;
                    }

                    if (cursor + 1 >= length && cursor + 1 <= inputList.size()) {
                        // cursor: 好, xSubPhrase: 各有所好
                        QStringView x_sub_phrase = cleanInput.mid(cursor + 1 - length, length);
                        auto it = d->phrases_dict2.find(x_sub_phrase);
                        if (it != d->phrases_dict2.end()) {
                            // overwrite pinyin
                            removeElements(result, cursor + 1 - length, length - 1);
                            result.append(it.value());
                            cursor += 1;
                            found = true;
                        }
                    }

                    if (found) {
                        break;
                    }

                    if (cursor + 2 >= length && cursor + 2 <= inputList.size()) {
                        // cursor: 好, xSubPhrase: 叶公好龙
                        QStringView x_sub_phrase_1 = cleanInput.mid(cursor + 2 - length, length);
                        auto it = d->phrases_dict2.find(x_sub_phrase_1);
                        if (it != d->phrases_dict2.end()) {
                            // overwrite pinyin
                            removeElements(result, cursor + 2 - length, length - 2);
                            result.append(it.value());
                            cursor += 2;
                            found = true;
                        }
                    }

                    if (found) {
                        break;
                    }
                }

                // not found, use default pinyin
                if (!found) {
                    result.append(d->getDefaultPinyin(current_char));
                    cursor++;
                }
            }
        }

        if (!tone) {
            for (QStringView &item : result) {
                if (item.back().isDigit()) {
                    item.chop(1);
                }
            }
        }

        // Alloc 2
        if (error == errorType::Ignore) {
            return joinView(result, QStringLiteral(" "));
        }
        return resetZH(input, result, inputPos);
    }

    ZhG2p::ZhG2p(ZhG2pPrivate &d, QObject *parent) : QObject(parent), d_ptr(&d) {
        d.q_ptr = this;

        d.init();
    }

    QString ZhG2p::tradToSim(const QString &text) const {
        Q_D(const ZhG2p);
        return d->tradToSim(text).toString();
    }

    bool ZhG2p::isPolyphonic(const QString &text) const {
        Q_D(const ZhG2p);
        return d->isPolyphonic(text);
    }
    QStringList ZhG2p::getDefaultPinyin(const QString &text) const {
        Q_D(const ZhG2p);
        return d->word_dict.value(d->tradToSim(text).toString(), {});
    }

}

File: src/libs/ikg2p/python/Zh_G2p.py
def is_letter(c):
    return ('a' <= c <= 'z') or ('A' <= c <= 'Z')


def is_hanzi(c):
    return 0x4e00 <= ord(c) <= 0x9fa5


def is_kana(c):
    return (0x3040 <= ord(c) <= 0x309F) or (0x30A0 <= ord(c) <= 0x30FF)


def is_special_kana(c):
    special_kana = "ャュョゃゅょァィゥェォぁぃぅぇぉ"
    return c in special_kana


class ZhG2p:
    def __init__(self, language):
        self.PhrasesMap = {}
        self.TransDict = {}
        self.WordDict = {}
        self.PhrasesDict = {}

        if language == "mandarin":
            dict_dir = "Dicts/mandarin"
        else:
            dict_dir = "Dicts/cantonese"

        self.load_dict(dict_dir, "phrases_map.txt", self.PhrasesMap)
        self.load_dict_list(dict_dir, "phrases_dict.txt", self.PhrasesDict)
        self.load_dict_list(dict_dir, "user_dict.txt", self.PhrasesDict)
        self.load_dict_list(dict_dir, "word.txt", self.WordDict)
        self.load_dict(dict_dir, "trans_word.txt", self.TransDict)

    @staticmethod
    def load_dict(_dict_dir, _file_name, _result_map):
        dict_path = _dict_dir + "/" + _file_name
        with open(dict_path, 'r', encoding='utf-8') as f:
            for line in f:
                line = line.strip()
                if line:
                    key, value = line.split(':')
                    _result_map[key] = value

    @staticmethod
    def load_dict_list(_dict_dir, _file_name, _result_map):
        dict_path = _dict_dir + "/" + _file_name
        with open(dict_path, 'r', encoding='utf-8') as f:
            for line in f:
                line = line.strip()
                if line:
                    key, value = line.split(':')
                    _result_map[key] = value.split(' ')

    NumMap = {
        "0": "零",
        "1": "一",
        "2": "二",
        "3": "三",
        "4": "四",
        "5": "五",
        "6": "六",
        "7": "七",
        "8": "八",
        "9": "九"
    }

    @staticmethod
    def reset_zh(_input, res, positions):
        result = _input.copy()
        for i, pos in enumerate(positions):
            result[pos] = res[i]
        return " ".join(result)

    @staticmethod
    def remove_elements(lst, start, n):
        if 0 <= start < len(lst) and n > 0:
            count_to_remove = min(n, len(lst) - start)
            del lst[start:start + count_to_remove]

    @staticmethod
    def split_string_no_re(input_str):
        res = []
        pos = 0

        while pos < len(input_str):
            current_char = input_str[pos]

            if is_letter(current_char):
                start = pos
                while pos < len(input_str) and is_letter(input_str[pos]):
                    pos += 1
                res.append(input_str[start:pos])
            elif is_hanzi(current_char) or current_char.isdigit():
                res.append(input_str[pos])
                pos += 1
            elif is_kana(current_char):
                length = 2 if pos + 1 < len(input_str) and is_special_kana(input_str[pos + 1]) else 1
                res.append(input_str[pos:pos + length])
                pos += length
            else:
                pos += 1

        return res

    def convert(self, _input, tone=False, convert_num=False):
        return self.convert_list(self.split_string_no_re(_input), tone, convert_num)

    def zh_position(self, _input, res, positions, convert_num):
        for i, val in enumerate(_input):
            if val in self.WordDict or val in self.TransDict:
                res.append(val)
                positions.append(i)
            elif convert_num and val in self.NumMap:
                res.append(self.NumMap[val])
                positions.append(i)

    def convert_list(self, _input, tone=False, convert_num=False):
        input_list = []
        input_pos = []
        self.zh_position(_input, input_list, input_pos, convert_num)
        clean_input = ''.join(input_list)
        result = []
        cursor = 0

        while cursor < len(input_list):
            raw_current_char = input_list[cursor]
            current_char = self.trad_to_sim(raw_current_char)

            if current_char not in self.WordDict:
                result.append(current_char)
                cursor += 1
                continue

            if not self.is_polyphonic(current_char):
                result.append(self.get_default_pinyin(current_char))
                cursor += 1
            else:
                found = False
                for length in range(4, 1, -1):
                    if cursor + length <= len(input_list):
                        sub_phrase = clean_input[cursor:cursor + length]
                        if sub_phrase in self.PhrasesDict:
                            result.extend(self.PhrasesDict[sub_phrase])
                            cursor += length
                            found = True

                        if cursor >= 1 and not found:
                            sub_phrase_1 = clean_input[cursor - 1:cursor + length - 1]
                            if sub_phrase_1 in self.PhrasesDict:
                                result = result[:-1]
                                result.extend(self.PhrasesDict[sub_phrase_1])
                                cursor += length - 1
                                found = True

                    if 0 <= cursor + 1 - length < cursor + 1 and not found and cursor < len(input_list):
                        x_sub_phrase = clean_input[cursor + 1 - length:cursor + 1]
                        if x_sub_phrase in self.PhrasesDict:
                            self.remove_elements(result, cursor + 1 - length, length - 1)
                            result.extend(self.PhrasesDict[x_sub_phrase])
                            cursor += 1
                            found = True

                    if 0 <= cursor + 2 - length < cursor + 2 and not found and cursor < len(input_list):
                        x_sub_phrase = clean_input[cursor + 2 - length:cursor + 2]
                        if x_sub_phrase in self.PhrasesDict:
                            self.remove_elements(result, cursor + 2 - length, length - 1)
                            result.extend(self.PhrasesDict[x_sub_phrase])
                            cursor += 2
                            found = True

                if not found:
                    result.append(self.get_default_pinyin(current_char))
                    cursor += 1

        if not tone:
            result = [x[:-1] if x[-1].isdigit() else x for x in result]

        return self.reset_zh(_input, result, input_pos)

    def is_polyphonic(self, text):
        return text in self.PhrasesMap

    def trad_to_sim(self, text):
        return self.TransDict[text] if text in self.TransDict else text

    def get_default_pinyin(self, text):
        return self.WordDict[text][0] if text in self.WordDict else None

    def get_all_pinyin(self, text):
        return self.WordDict[text] if text in self.WordDict else None